### DISCUSSION TITLE
Can't establish a connection with SecurityPolice

### DISCUSSION THREAD

**Discussion by fkemper** [locka99/opcua#246] [Q&A] [answered]
Hello,

I'm new to opcua and rust. i have problems to establishe a connection to a plc-opc-server with SecurityPolice other than NONE.

I use the sample "simple-client".

What is working:

Connect to the plc-opc -server with the following client-config.

```
let mut client = ClientBuilder::new()
.application_name("Simple Client")
.application_uri("urn:SimpleClient")
.product_uri("urn:SimpleClient")
.trust_server_certs(true)
.create_sample_keypair(true)
.session_retry_limit(3)
.client()
.unwrap();
```
and end-point-config

```
...
if let Ok(session) = client.connect_to_endpoint(
    (
        args.url.as_ref(),
       
        SecurityPolicy::None.to_str(),
        MessageSecurityMode::None,
        
        UserTokenPolicy::anonymous(),
    ),
    IdentityToken::UserName(String::from("john"),String::from("doo")),
)  
...
```
![image](https://user-images.githubusercontent.com/22469707/207629270-9e57f1d5-1aa8-47a2-9332-73ac3fe165be.png)

What is **NOT** working:

end-point-config
```
if let Ok(session) = client.connect_to_endpoint(
    (
        args.url.as_ref(),
        /*
        SecurityPolicy::None.to_str(),
        MessageSecurityMode::None,
        */
        SecurityPolicy::Basic256Sha256.to_str(),
        MessageSecurityMode::SignAndEncrypt,
        
        UserTokenPolicy::anonymous(),
    ),
    IdentityToken::UserName(String::from("john"),String::from("doo")),
)  
```

i get the following error-message

![image](https://user-images.githubusercontent.com/22469707/207630485-308a8ade-2971-4163-846f-e9394f57bc38.png)

But with another client i can connect to the server with this security-policy

![image](https://user-images.githubusercontent.com/22469707/207631015-100a35f7-0f02-4a2c-9b7d-da855dfc0a3d.png)
![image](https://user-images.githubusercontent.com/22469707/207634562-668c5ca4-33de-4582-9be7-2838a4ea5939.png)

What thing i do make wrong? Why the other client can connect to the server?

Thanks in advance

**Comment by schroeder-** [comment, marked as answer]
As the log warns about. The server certificate is not valid. To skip the validation add "verify_server_certs(false)" to the ClientBuilder.
``` Rust
let mut client = ClientBuilder::new()
.application_name("Simple Client")
.application_uri("urn:SimpleClient")
.product_uri("urn:SimpleClient")
.trust_server_certs(true)
.create_sample_keypair(true)
.session_retry_limit(3)
.verify_server_certs(false)
.client()
.unwrap();
```

**Comment by fkemper** [comment]
Hello,

you safe my day...

i insert the `.verify_server_certs(false)` and now the connection be establish

But where do you read that the certificate is not valid? I only read that the certificate was rejected.  The reason for the rejection was not clear for me...

I research something why the server-certivicate was not valid and i found something in the config-file of the plc-opc-server...

So it seeem that the other client doesn't make a validation of the certificate...

**Reply by schroeder-** [reply]
You have to lower the loglevel to output warnings, to get the reason for the rejection.

**Reply by fkemper** [reply]
Thanks for the tip
